core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -62,14 +64,70 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
   def fromSeq[A](as: Seq[A]): Option[NonEmptyLazyList[A]] =
     if (as.nonEmpty) Option(create(LazyList.from(as))) else None
 
-  def fromLazyListPrepend[A](a: A, ca: LazyList[A]): NonEmptyLazyList[A] =
-    create(a +: ca)
+  def fromLazyListPrepend[A](a: => A, ll: => LazyList[A]): NonEmptyLazyList[A] =
+    create(a #:: ll)
+
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListPrepend[A]()(a: A, ll: LazyList[A]): NonEmptyLazyList[A] =
+    fromLazyListPrepend(a, ll)
+
+  def fromLazyListAppend[A](ll: => LazyList[A], a: => A): NonEmptyLazyList[A] =
+    create(ll #::: a #:: LazyList.empty)
 
-  def fromLazyListAppend[A](ca: LazyList[A], a: A): NonEmptyLazyList[A] =
-    create(ca :+ a)
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListAppend[A]()(ll: LazyList[A], a: A): NonEmptyLazyList[A] =
+    fromLazyListAppend(ll, a)
 
   def apply[A](a: => A, as: A*): NonEmptyLazyList[A] =
-    create(LazyList.concat(LazyList(a), LazyList.from(as)))
+    create(a #:: LazyList.from(as))
+
+  /**
+   * Wraps a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  def maybe[A](ll: => LazyList[A]): Maybe[A] = new Maybe(() => ll)
+
+  /**
+   * Wrapper around a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  final class Maybe[A] private[NonEmptyLazyList] (mkLL: () => LazyList[A]) {
+    // because instances of this class are created explicitly, they might be
+    // reused, and we don't want to re-evaluate `mkLL`
+    private[this] lazy val ll = mkLL()
+
+    /** Prepends a single element, yielding a `NonEmptyLazyList`. */
+    def #::[AA >: A](elem: => AA): NonEmptyLazyList[AA] =
+      create(elem #:: ll)
+
+    /** Prepends a `LazyList`, yielding another [[Maybe]]. */
+    def #:::[AA >: A](prefix: => LazyList[AA]): Maybe[AA] =
+      new Maybe(() => prefix #::: ll)
+
+    /** Prepends a `NonEmptyLazyList`, yielding a `NonEmptyLazyList`. */
+    def #:::[AA >: A](prefix: => NonEmptyLazyList[AA])(implicit d: DummyImplicit): NonEmptyLazyList[AA] =
+      create(prefix.toLazyList #::: ll)
+  }
+
+  final class Deferrer[A] private[NonEmptyLazyList] (private val nell: () => NonEmptyLazyList[A]) extends AnyVal {
+
+    /** Prepends a single element. */
+    def #::[AA >: A](elem: => AA): NonEmptyLazyList[AA] =
+      create(elem #:: nell().toLazyList)
+
+    /** Prepends a `LazyList`. */
+    def #:::[AA >: A](prefix: => LazyList[AA]): NonEmptyLazyList[AA] =
+      create(prefix #::: nell().toLazyList)
```

Comment:
The similar method from `LazyList.Deferrer` does not use by-name parameter here:
https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/LazyList.scala#L1144-L1147
```scala
    /** Construct a LazyList consisting of the concatenation of the given LazyList and
      *  another LazyList.
      */
    def #:::[B >: A](prefix: LazyList[B]): LazyList[B] = prefix lazyAppendedAll l()
```
So it looks like this `Deferrer` pretends to be more lazy than one from `LazyList`. Is it really necessary here?

Reply:
oops, good catch. it's been a while since I touched `LazyList`, and I forgot which side is lazy. (also I kind of just went through and made all the parameters by-name without particularly thinking about it)

Reply:
I'm not actually sure why it isn't by-name in `LazyList`—it doesn't seem like it would be overly difficult to implement, and seems like a straight win from a laziness perspective

Reply:
> So it looks like this `Deferrer` pretends to be more lazy than one from `LazyList`

I used the `LazyList.empty #:::` trick to make it actually more lazy and no longer pretending

Reply:
>  and seems like a straight win from a laziness perspective

Well, seems it uses `lazyAppendedAll` under the hood which is called on `this` and takes a by-name `suffix` param. Therefore deferring `this` does not seem providing any benefit since it will be forced anyway as soon as it is accessed.

Reply:
> Well, seems it uses `lazyAppendedAll` under the hood

yes but, to use the `LazyList` internal APIs (you asked for it 😜), you could implement it with a by-name prefix and better laziness as follows:

```scala
def #:::[B >: A](prefix: => LazyList[B]): LazyList[B] = newLL(prefix.state) lazyAppendedAll l()
```

I'll take the blame for missing it, as it was primarily my baby

Reply:
side note: there's also a fairly obvious hole in the API which it's too late to fix of a `deferred`/`defer` method that takes a by-name parameter and makes its evaluation lazy. but again, too late to fix :(

perhaps someday there will be a 2.14 or equivalent and I can fix it...

Reply:
> perhaps someday there will be a 2.14 or equivalent and I can fix it...

Once Scala 2.13 relaxes its forward-compatibility as of https://github.com/scala/improvement-proposals/pull/54, we can fix it :)

Reply:
@armanbilge thanks! I'll have to keep track of that (or ask someone to poke me when it happens), and perhaps that'll get me back into stdlib work. who knows!